src/common/mfu_flist_copy.c

```diff
@@ -1554,50 +1587,91 @@ static int mfu_is_eof(const char* file, int fd)
     uint64_t length,
     uint64_t file_size,
-    mfu_copy_opts_t* mfu_copy_opts)
+    mfu_copy_opts_t* mfu_copy_opts,
+    mfu_file_t* mfu_src_file,
+    mfu_file_t* mfu_dst_file)
 {
-    /* hint that we'll read from file sequentially */
-//    posix_fadvise(in_fd, offset, chunk_size, POSIX_FADV_SEQUENTIAL);
-
+    /* There is no equivalent "lseek" in daos, it is part of read, so skip that
+     * for a DAOS path */
     /* seek to offset in source file */
-    if(mfu_lseek(src, in_fd, offset, SEEK_SET) == (off_t)-1) {
-        MFU_LOG(MFU_LOG_ERR, "Couldn't seek in source path `%s' (errno=%d %s)",
-            src, errno, strerror(errno));
+    if(mfu_file_lseek(src, mfu_src_file, offset, SEEK_SET) == (off_t)-1) {
+        MFU_LOG(MFU_LOG_ERR, "Couldn't seek in source path `%s' (errno=%d %s)", src, errno, strerror(errno));
         return -1;
     }
 
     /* seek to offset in destination file */
-    if(mfu_lseek(dest, out_fd, offset, SEEK_SET) == (off_t)-1) {
-        MFU_LOG(MFU_LOG_ERR, "Couldn't seek in destination path `%s' (errno=%d %s)",
-            dest, errno, strerror(errno));
+    if (mfu_file_lseek(dest, mfu_dst_file, offset, SEEK_SET) == (off_t)-1) {
+        MFU_LOG(MFU_LOG_ERR, "Couldn't seek in destination path `%s' (errno=%d %s)", dest, errno, strerror(errno));
         return -1;
     }
 
-    /* get buffer */
+    /* declare daos vars */
+#ifdef DAOS_SUPPORT
+    daos_size_t got_size;
+    d_sg_list_t sgl;
+    d_iov_t     iov;
+#endif
+
+    /* set buffer and buffer size */
     size_t buf_size = mfu_copy_opts->block_size;
     void* buf = mfu_copy_opts->block_buf1;
+    void* newbuf;
+
+    /* set daos vars and buffers */
+#ifdef DAOS_SUPPORT
+    if (mfu_src_file->type == DAOS || mfu_dst_file->type == DAOS) {
+        sgl.sg_nr = 1;
+        D_ALLOC(buf, buf_size);
+        D_ASSERT(buf != NULL);
+        d_iov_set(&iov, buf, buf_size);
+        sgl.sg_iovs = &iov;
+        sgl.sg_nr_out = 1;
+        if (mfu_src_file->type == DAOS) {
+            mfu_src_file->offset = (daos_off_t) offset;
+            mfu_src_file->sgl = &sgl;
+        }
+        if (mfu_dst_file->type == DAOS) {
+            mfu_dst_file->offset = (daos_off_t) offset;
+            mfu_dst_file->sgl = &sgl;
+        }
+    }
+#endif
 
     /* write data */
     size_t total_bytes = 0;
-    while(total_bytes < (size_t)length) {
+    char* buffer = mfu_copy_opts->block_buf1;
+    while(total_bytes <= (size_t)length) {
         /* determine number of bytes that we
          * can read = max(buf size, remaining chunk) */
         size_t left_to_read = (size_t)length - total_bytes;
+        
         if(left_to_read > buf_size) {
+#ifdef DAOS_SUPPORT
+            if (mfu_src_file->type == DAOS || mfu_dst_file->type == DAOS) {
+                newbuf = mfu_copy_opts->block_buf1;
+  	        D_REALLOC(newbuf, buf, left_to_read);
+                d_iov_set(&iov, newbuf, left_to_read);
+            }
+#endif
             left_to_read = buf_size;
         }
```

Comment:
Something here feels a bit off to me, but I might just have the wrong interpretation.  This if branch limits the buffer size to be no bigger than buf_size.  For example, coming in left_to_read might be 1GB but buf_size might only be 1MB.  This will reduce left_to_read down to 1MB, so that we only read from the file in chunks up to 1MB at a time.  However, the DAOS code is using the left_to_read value before it is reduced.  Should this code be moved closer to the read operation, after all of the manipulation to left_to_read has been completed?

Reply:
@adammoody this should be resizing the DAOS buffer in the scatter gather list, which is used to tell DAOS where the data is in the DAOS read call below this one. I don't know what you mean about DAOS using the left_to_read before it is reduced?

Reply:
@adammoody never mind, after giving this another look I see that resizing the DAOS buffer in the io vector is unnecessary. I am taking out that block.

Reply:
The left_to_read can change values a few times before finally reaching the mfu_file_read the call.  So if buf_size=1MB, it might do something like:
```
// depending on the file size, this might be something really big like left_to_read=100GB
// or it might be something small like left_to_read=10 bytes
left_to_read = length - total_bytes;

// in case left_to_read is really big, this reduces it to be at most buf_size
if (left_to_read > buf_size)
  left_to_read = buf_size; // reduce left_to_read to 1MB

// this check is for cases where left_to_read < buf_size,
// because O_DIRECT requires particular values for the read count
if (synchronous)
  left_to_read = buf_size; // force left_to_read to be exactly 1MB
```

Reply:
@adammoody yeah, thanks I appreciate it, I don't actually need to resize the io vector buffer there.

Reply:
Does it work if you change this:
```
        if(left_to_read > buf_size) {
#ifdef DAOS_SUPPORT
            if (mfu_src_file->type == DAOS || mfu_dst_file->type == DAOS) {
                newbuf = mfu_copy_opts->block_buf1;
  	        D_REALLOC(newbuf, buf, left_to_read);
                d_iov_set(&iov, newbuf, left_to_read);
            }
#endif
            left_to_read = buf_size;
        }
```
to just:
```
        if(left_to_read > buf_size) {
            left_to_read = buf_size;
        }
```
and then add a d_iov_set line here:
```
#ifdef DAOS_SUPPORT
        if (mfu_src_file->type == DAOS || mfu_dst_file->type == DAOS) {
            d_iov_set(&iov, buf, left_to_read);
            buffer = sgl.sg_iovs[0].iov_buf;
        }
#endif
        num_of_bytes_read = mfu_file_read(src, buffer, left_to_read, mfu_src_file);
```

Reply:
Note that I'm also dropping the D_REALLOC, which I'm guessing might not be necessary.  Above the while loop, buf is allocated to be buf_size bytes, and we never read more than that amount at a time.  We might read less, but I think it'd be fine to use a buffer that is bigger than is needed, so we don't have to allocate a new buffer.  We just need to set the size correctly on the scatter/gather list element.

Reply:
@adammoody well I don't actually need to resize the buffer to left_to_read at all, since it can just read buf_size each time, which is fine. That was just a mistake on my part I think.